Feat: add new docker for production & development environment

toofff opened the pull request:
- [x] add new `docker-compose.yml` & `Dockerfile` layer divisions
    * [x] service `php` & configuration (env PROD & DEV)
        * [x] use php:74-fpm-alpine
        * [x] add docker-healthcheck
        * [x] update docker-entrypoint
            * [x] use php.ini PROD or DEV
            * [x] create important folders for the use of the project (`public/files`, `public/theme`, `public/thumbs`, `var/cache`, `var/log` & `var/sessions`)
            * [x] install vendor for development environment
            * [x] test BDD is active
            * [x] use `php bin/console bolt:info` instead of launching after installing vendors (because it is looking for the version of the database which might not be launched in time)
            * [x] if the theme does not exist, it will copy it
            * [x] if this project use DoctrineMigrationBundle then launch migrations (all env) else reset & create schema (only env DEV & TEST)
            * [x] launch fixtures
    * [x] service `nginx` & configuration (env PROD & DEV)
        * [x] copy public folder of php container
    * [x] service `db` with MySQL 5.7 version (only env DEV)
    * [x] service `h2-proxy` to have SSL (only env DEV)
    * [x] service `mailcatcher` to do (only env DEV)
- [x] add `.dockerignore` file
- [x] add `.editorconfig` file
- [x] replace `.env.dist` by `.env` file, to respect symfony's good practice
- [x] rework the `.gitignore` file
- [x] rework the `Makefile` file
- [x] rework the `README.md` file
- [x] rework the `composer.json` file & update dependencies
    * [x] `bolt/core` 4.0.1
    * [x] `bolt/themes` 3.3.2
- [x] update the `symfony.lock` file
- [x] fix the use of API-Platform
    * [x] the configuration (add path `%kernel.project_dir%/src/Entity` for the use of our own entity)
    * [x] install assets with the copy file
- [x] add dev dependency for ECS & PHPStan
- [x] upgrade documentation
- [x] rework commit in pull request


I added a CI for the installation of a Bolt project with Composer or Docker Compose.

- [x] Bolt project installation test with docker-compose, PHP 7.4 and MySQL
    - [x] Validate composer.json
    - [x] Run PHP-CS-Fixer
    - [x] Run PHP-CS-Fixer with fix flag
    - [x] Run PHPStan
    - [x] Run Security Checker
    - [x] Check HTTP reachability
        - [x] http://localhost:8080
        - [x] http://localhost:8080/bolt
        - [x] http://localhost:8080/api/contents
        - [x] http://localhost:8080/api/fields
        - [x] http://localhost:8080/api/relations
        - [x] https://localhost:8443
        - [x] https://localhost:8443/bolt
        - [x] https://localhost:8443/api/contents
        - [x] https://localhost:8443/api/fields
        - [x] https://localhost:8443/api/relations
- [x] Bolt project installation test with composer, PHP 7.4 and MySQL
    - [x] Validate composer.json
    - [x] Run PHP-CS-Fixer
    - [x] Run PHP-CS-Fixer with fix flag
    - [x] Run PHPStan
    - [x] Run Security Checker
    - [x] Check HTTP reachability
        - [x] http://localhost:8000
        - [x] http://localhost:8000/bolt
        - [x] http://localhost:8000/api/contents
        - [x] http://localhost:8000/api/fields
        - [x] http://localhost:8000/api/relations

The continuous integration is not yet working on this pull request because nothing is merged into the Bolt repository yet in my opinion. On the other hand you can see it working on [my pull request in my fork](https://github.com/toofff/bolt-project/pull/2/checks)

bobdenotter reviewed (commented):
I don't use Docker myself, but I know a lot of people do, so it's great if this is updated thoroughly. :-)

Some questions about the first things I noticed, but you might still be working on this: 

 - I see you've _removed_ a few files (`.gitignore`'s, `.env.dist`). Is this on purpose? 
 - We're keeping our version of PHP in sync with Symfony. So, we'll need to make sure the `7.4` in `symfony.lock` won't break that requirement.

toofff commented:
- Yes I deleted the `.gitignore` file in the public folder. I prefer to have the project rules at the root but above all it forces you to have server rules so that this file is not accessible in production if it is forgotten during deployment.

- Yes I deleted the `.env.dist` file in the root project. [Here are the new specs on this file](https://symfony.com/doc/current/configuration/dot-env-changes.html) since `Symfony / flex` and `Symfony 5` 

- For PHP 7.4 it is a proposal, I can downgrade if you wish? Even though I personally prefer to be on the last minor version of the 7.x branch for creating a new project.

I'm glad you welcome this work like this. I still have some work to do but you can already test it with a simple `docker-compose up`. Everything is functional at home.

If you have more proposals, I am interested ;)

toofff commented:
@bobdenotter 

Little question, do you know why we had a "node" container in the docker?

Great news, I think I finished it all this week. Don't hesitate to test and give feedback ;)

bobdenotter commented:
> Little question, do you know why we had a "node" container in the docker?

No, i'm afraid I don't.. I don't really use Docker myself. 

Perhaps @JarJak (🛎️) knows?

toofff commented:
@bobdenotter I still have the documentation to finish, I hope to deliver it to you this week for the review :smile:

bobdenotter commented:
@toofff Excellent! Looking forward to it!

toofff commented:
@bobdenotter Now I pass the torch to you for proofreading and testing locally. I remain at your disposal for any questions and changes to be made. :)

bobdenotter commented:
@toofff Thanks! 

It'll take me a little while to review, because I'd like to do it with care. Will let you know if any questions arise.

alanpoulain commented on .editorconfig:
```diff
@@ -0,0 +1,88 @@
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.feature]
```
Why not gathering the files with the same style?

toofff replied:
to make it more understandable

alanpoulain replied:
Why not, it's subjective anyway :slightly_smiling_face:

alanpoulain commented on README.md:
```diff
@@ -48,17 +61,89 @@ symfony server:start -d
+Start by [downloading the Bolt project distribution `.tar.gz` file](https://github.com/bolt/project/releases/latest), or [generate a GitHub repository from the template we provide](https://github.com/bolt/project/generate).
+Once you have extracted its contents, the resulting directory contains the Bolt project structure. You will add your own code and configuration inside it.
+
+**Note**: Try to avoid using the `.zip` file, as it may cause potential permission.
```
```suggestion
**Note**: Try to avoid using the `.zip` file, as it may cause potential permission issues.
```

alanpoulain commented on README.md:
```diff
@@ -48,17 +61,89 @@ symfony server:start -d
+### with Docker
+
+Start by [downloading the Bolt project distribution `.tar.gz` file](https://github.com/bolt/project/releases/latest), or [generate a GitHub repository from the template we provide](https://github.com/bolt/project/generate).
+Once you have extracted its contents, the resulting directory contains the Bolt project structure. You will add your own code and configuration inside it.
```
```suggestion
Once you have extracted its content, the resulting directory contains the Bolt project structure. You will add your own code and configuration inside it.
```

alanpoulain commented on README.md:
```diff
@@ -48,17 +61,89 @@ symfony server:start -d
+
+Open a terminal, and navigate to the directory containing your project skeleton.
+
+Create the `docker-compose.override.yml` file at the root project for development environment and added the following:
```
```suggestion
Create the `docker-compose.override.yml` file at the root project for development environment and add the following:
```

alanpoulain commented on README.md:
```diff
@@ -48,17 +61,89 @@ symfony server:start -d
+    environment:
+      - RESET_DATABASE=true  # create schema in database or reset database if the value is "true"
+      - LAUNCH_FIXTURES=true # load fixtures in databse if the value is "true"
+      # create admin user for development environment (attention fixtures already created an admin user)
```
```suggestion
      # create admin user for development environment (warning: fixtures already created an admin user)
```

alanpoulain commented on README.md:
```diff
@@ -48,17 +61,89 @@ symfony server:start -d
+
+  docker-compose exec php sh # To enter the container directly, you will be placed at the root of the project
+
+  docker-compose exec php bin/concole bolt:add-user # follow the creation steps in the console (attention fixtures already created an admin user)
```
```suggestion
  docker-compose exec php bin/concole bolt:add-user # follow the creation steps in the console (warning: fixtures already created an admin user)
```

alanpoulain commented on README.md:
````diff
@@ -48,17 +61,89 @@ symfony server:start -d
+  <summary>To see the container's logs, run:</summary>
+
+  ```bash
+  docker-compose logs        # display the logs of the all containers
````
```suggestion
  docker-compose logs        # display the logs of all containers
```

alanpoulain commented on docker/php/docker-entrypoint.sh:
```diff
@@ -0,0 +1,65 @@
+fi
+
+if [ "$1" = 'php-fpm' ] || [ "$1" = 'php' ] || [ "$1" = 'bin/console' ]; then
+	# todo: a vérifier dans le container et l'image.
```
Comment in French :smile:

toofff replied:
Ooops, my bad :(

toofff commented on docker/php/docker-entrypoint.sh:
```diff
@@ -0,0 +1,65 @@
+		composer run post-create-project-cmd
+	fi
+
+	if ls -A migrations/*.php > /dev/null 2>&1; then
```
This part I do not like it too much but I have not found better. I prefer to explain my reasoning to you.

For the construction of the database I am in favor of [doctrine migrations](https://symfony.com/doc/master/bundles/DoctrineMigrationsBundle/index.html) for all environments and especially for tests and production.
```
	if ls -A migrations/*.php > /dev/null 2>&1; then
		bin/console doctrine:migrations:migrate --no-interaction
	fi
```
Do you want me to add depending on the project?
I can add it in the documentation, but where?

Migrations are cool but when we start the project we don't have any yet.
To overcome this problem and only on dev and test environments, I implemented this:
1. Reset the database with the envrioment variable `RESET_DATABASE`
2. Launch of bolt fixtures with the envrioment variable `LAUNCH_FIXTURES`
3. The create admin user with envrioment variables `ADMIN_USERNAME`, `ADMIN_PASSWORD` & `ADMIN_EMAIL`
```
	if [ "$APP_ENV" != 'prod' ]; then
		if [ ! -z $RESET_DATABASE ]; then
			bin/console doctrine:schema:drop --force --full-database
			bin/console doctrine:schema:create
		fi

		if [ ! -z $LAUNCH_FIXTURES ]; then
			bin/console doctrine:fixtures:load --no-interaction
		fi

		if [ ! -z $ADMIN_USERNAME ] && [ ! -z $ADMIN_PASSWORD ] && [ ! -z $ADMIN_EMAIL ]; then
			bin/console bolt:add-user $ADMIN_USERNAME $ADMIN_PASSWORD $ADMIN_EMAIL $ADMIN_USERNAME --admin --no-interaction
		fi
	fi
```

This is documented in the Readme file in the Docker part when creating the `docker-entrypoint.override.yml` file.
This forces the person to read the documentation carefully otherwise the project could potentially not work if it does not use one of the 2 solutions for the creation of the database (create `doctrine migrations` or add `RESET_DATABASE` at `true` in the `docker-entrypoint.override.yml` file)

What do you think @bobdenotter, @alanpoulain ?

alanpoulain replied:
Personally, I wouldn't like the database to be migrated / reset each time I start the containers.
I think this entrypoint shouldn't do anything if there is no `docker-entrypoint.override.yml`.
The migrations should also be applied only if the user has explicitly tells so IMO.

toofff replied:
What I did with the different conditions set up. I leave the choice to the user to use it or not ;)

simongroenewolt replied:
> Personally, I wouldn't like the database to be migrated / reset each time I start the containers.
> I think this entrypoint shouldn't do anything if there is no `docker-entrypoint.override.yml`.
> The migrations should also be applied only if the user has explicitly tells so IMO.

Same here. I'm using docker a lot and prefer the start-up of the container(s) to not do anything that isn't necessary. If the default is 'do nothing' I'd be completely ok with extra features, but watch out for maintenance!

toofff replied:
@simongroenewolt

What I wanted to do is to have a Bolt project that works as soon as we start it without additional commands.
We need a viable database for Bolt to be functional at startup without further intervention. That's why I proposed this system there.

So I have proposed several solution that starts when we tell them explicitly.

If the Doctrine miagrations folder is present, then it builds the database with that.
If we have nothing, I leave the possibility to the developer:
- either build the database with the env variable `$RESET_DATABASE`
- either inject the Bolt fixtures into the database with the env variable `$LAUNCH_FIXTURES`
- either add a new administrator in the database with the env variables `$ADMIN_USERNAME`, `$ADMIN_PASSWORD` & `$ADMIN_EMAIL`

I grant you, it's very practical when testing Bolt and having to start from scratch or when you initialize the project the first time. For the long term of course the developer will do what he wants.

After that, this is only a proposal to prevent the developer from placing additional orders.
I can remove all of that and document the commands to run directly with Docker if you mind?

simongroenewolt replied:
@toofff I understand your goal of wanting to get the developer up and running with minimal interaction. I know that bolt/core had a migration in the repository once, but Bob told me that didn't belong there as the migrations are db specific (mysql/sqlite/postgres).
So I'm not really expecting bolt/platform to supply migrations in the package download either. If the only migrations that are expected will have been generated by the developer they don't need to be auto-run at start do they? Or am I missing something?

toofff replied:
If the user chooses to use doctrine migrations, then they must be launched from the container run, for all environments, otherwise the project will launch and will fall into error because the database will not be up to date depending on the entities set up.

The best practice when using doctrine migrations is to create the test database with it instead of doing a `doctrine:schema:update --force` or `bin/console doctrine:schema:create`. So we indirectly test our doctrines migrations by launching our various tests. :smile:

toofff replied:
Since this part was a bit controversial (even for me), I went back to it and removed the addition of a `docker-compose.override.yml` file. Obviously the commands launched automatically are replaced by good documentation in the `README.md` file.

chalasr commented on public/.htaccess:
```diff
@@ -1,69 +0,0 @@
-# Use the front controller as index file. It serves as a fallback solution when
```
Same here, apache configuration seems to be maintained on purpose.

simongroenewolt commented:
> 
> Little question, do you know why we had a "node" container in the docker?
> 

I don't know if you still have this question, but I think the node container is to build/process the javascript for the (vue based) editor support in /bolt

toofff commented:
Thanks @simongroenewolt I'm watching this this week.

toofff commented:
@simongroenewolt @bobdenotter 

For me the container node is not necessary here. From what I could understand, this container is only needed for "bolt/assets" development.

toofff commented:
@bobdenotter did you have time to test with and without docker?

toofff commented on docker-compose.yml:
```diff
@@ -1,69 +1,69 @@
+      - ./public:/srv/bolt/public:ro
+    ports:
+      - target: 80
+        published: 80
```
Even if I am against this change, because for me it is better to stay iso in production.
But hey, I'm still going to make this change :)

toofff replied:
Currently with the docker in place, Bolt runs on port 80 and port 443 (for the part with SSL).

Ok for the change.

But to have more consistency with its 2 ports, I suggest port 80 in 8080 and port 443 in 8443. Is it good for you?

The change will be made in the readme file and also in the documentation I am preparing.

simongroenewolt commented on docker-compose.yml:
```diff
@@ -1,69 +1,69 @@
+      - MYSQL_DATABASE=bolt
+      - MYSQL_USER=bolt
+      - MYSQL_PASSWORD=!ChangeMe!
+      - MYSQL_ROOT_PASSWORD=!ChangeMe!
```
I'm not sure if that makes a difference for credential exposure. If I'm interpreting https://symfony.com/doc/current/configuration.html#overriding-environment-values-via-env-local correctly .env should be checked in nowadays.

toofff replied:
I don't see what this changes with your solution, the `.env` file will be versioned in any case. Or otherwise I don't understand what you mean.

For me the docker-compose.yml file is really only to be used for development od the project or to run the tests in a CI.

In any case, when going into production without docker, we will simply not send the `docker` folder and the `docker-compose.yml` file to production. So changing places its variables is not necessary.

If you use docker images in production, it is not advisable to use a database image in production but a managed service.

And if you want to add the variables in the .env already in place, it is dedicated to the Symfony application and it would be a shame to add variables which are only used for development and only used for docker-compose.

And even if it is not advisable to use docker-compose in production, it is still possible. In that event best practices and to have a dedicated docker-compose.prod.yml file for production.

JKetelaar replied:
I think what @djboboch is trying to say is to add to your `.env` something like:
```
DATABASE_HOST=db
DATABASE_USERNAME=bolt
DATABASE_PASSWORD=bolt
DATABASE_NAME=bolt

DATABASE_URL=mysql://${DATABASE_USERNAME}:${DATABASE_PASSWORD}@${DATABASE_HOST}:3306/${DATABASE_NAME}
```

Then in your `docker-compose.yml` you can do:
```
    environment:
      MYSQL_DATABASE: ${DATABASE_NAME}
      MYSQL_USER: ${DATABASE_USERNAME}
      MYSQL_PASSWORD: ${DATABASE_PASSWORD}
      MYSQL_RANDOM_ROOT_PASSWORD: 'yes'
```

Which then allows you to automatically set up a dev environment without setting your .env.local

JKetelaar replied:
And if you want to avoid issue #36, you could manually set a env file specifically for Docker using `env_file` in docker-compose.yml

toofff replied:
Yes that's what I understood @JKetelaar.

For me in a Symfony project, the `.env` file should be used for the Symfony project.

```
DATABASE_USERNAME = bolt
DATABASE_PASSWORD = bolt
DATABASE_NAME = bolt
```
It's 3 variables are just used for the development environment with the `docker-compose.yml` file.
Good practice is to use docker-compose only locally and especially not on a production server.

So the `.env` file as I wrote:
- is not polluted by Docker Compose variables (local environment) when you want to use Composer.
- is not polluted by Docker Compose variables (local environment) when you want to put into production
- is easy to modify if you want to use SQLLite (so no need to remove the variables you want me to add)

Everything is not good to put everything in variable, otherwise it becomes a mess in the project and you bring bugs which should not exist.

toofff replied:
If you put all your environment variables in the `env` file, there may be security issues.

Varibles for Docker Compose that will be sent to the Symfony application and vice versa.

The best is to have 2 separate files or an `.env` file for the Symfony application and declare the docker environment variables in the `docker-compose.yml` file (solution I chose).

JKetelaar replied:
Oh yeah, I agree, don't get me wrong, just made pointed out what I think this person meant!

toofff replied:
Thanks to you, at least I was able to be clearer in my words

bobdenotter commented:
@toofff Is this 'good to go' for now? If so, we'll merge it in shortly, and we fine-tune it before release of a new version of this repo.

toofff commented:
I try to finalize tomorrow and do my rebase so that it is mergeable

toofff commented:
@bobdenotter It's good for me, I made the last corrections of the CI, the rebase of the master branch, the update of composer 2.

You can test by following the `README.md` file and merge if you like.

If you want to view the CI, [it is available on my fork](https://github.com/toofff/bolt-project/pull/2/checks)

JKetelaar commented on docker/nginx/conf.d/default.conf:
```diff
@@ -0,0 +1,119 @@
+    # We use two location blocks here, the first is an exact match to the dashboard
+    # the next is a strict forward match for URIs under the dashboard. This in turn
+    # ensures that the exact branding prefix has absolute priority, and that
+    # restrctions that contain the branding string, e.g. "bolt.db", still apply.
```
restrictions

JKetelaar commented on public/index.php:
```diff
@@ -3,38 +3,29 @@
 declare(strict_types=1);
```
This was removed from the latest Symfony

toofff replied:
Good practice is to leave it on, at least for projects.

Symfony removed it for a simple reason, made it easier to maintain the Symfony framework for versions lower than php7.

https://github.com/symfony/recipes/pull/507#discussion_r259587271
https://github.com/symfony/symfony/issues/28423#issuecomment-466927534

JKetelaar commented on public/index.php:
```diff
@@ -3,38 +3,29 @@
-    }
-    (new Dotenv())->load(__DIR__.'/../.env');
-}
+require dirname(__DIR__) . '/vendor/autoload.php';
```
Any reason for using a different coding style than the Symfony ones? There shouldn't be spaces between concatenation within Symfony guidelines. Might wanna check on that for the entire PR

JKetelaar commented on public/index.php:
```diff
@@ -3,38 +3,29 @@
-    }
-    (new Dotenv())->load(__DIR__.'/../.env');
-}
+require dirname(__DIR__) . '/vendor/autoload.php';
```
Any reason for using a different coding style than the Symfony ones? There shouldn't be spaces between concatenation within Symfony guidelines. Might wanna check on that for the entire PR.

JKetelaar replied:
https://symfony.com/doc/current/contributing/code/standards.html#structure
>Add a single space around binary operators (==, &&, …), with the exception of the concatenation (.) operator;

toofff replied:
One particular reason, no. I use the ECS configuration of the project already set up.

I will look to correct this immediately.

JKetelaar reviewed (approved):
This has my blessing

bobdenotter reviewed (approved):
As previously discussed: Let's roll with this, and iron out kinks as we go! 

🚂 🚋 🚋
